File: core/kotlin/FanoutChaos.kt
// FanoutChaos.kt — RFC 0011 deterministic chaos engine, JVM/Kotlin port.
//
// "chaos contract" — mirrors core/c/fanout_chaos.c (the reference oracle)
// micro-step for micro-step, draw for draw. Given the same config, the
// stepped verdict JSON is BYTE-IDENTICAL to the C engine's (verified by
// ci/scripts/run_chaos_parity.sh; the golden fixture lives in
// tools/chaos-fixtures/). Read core/c/fanout_chaos.h for the normative
// contract; this port adds nothing and drops nothing.
//
// u32 discipline: Kotlin's UInt is the exact analog of C's uint32_t —
// times/shl wrap modulo 2^32, ushr is the logical right shift. No masking
// gymnastics needed; the contract vectors pin the behavior.
//
// LAW 4 honesty: the stepped engine models the RFC 0004 protocol exactly
// (FI1/FI2 brackets, the bounded 4-attempt claim loop, latest-wins). The
// real-ring JVM chaos leg (threads + chaos points over WeftFanoutBroadcaster)
// lives in the Android test suite; this file is the parity oracle.

import kotlin.system.exitProcess

data class ChaosConfig(
    val seed: Int,        // master seed (u32)
    val steps: Long,      // scheduler step budget (stepped mode)
    val slots: Int,       // M — ring depth
    val words: Int,       // W — payload words per slot
    val readers: Int,     // R — concurrent reader SMs
    val frames: Int,      // F — frames the writer publishes
    val chaosRate: Int,   // per-mille fault probability per step
)

data class ChaosVerdict(val pass: Boolean, val engine: String, val json: String)

// ---------------------------------------------------------------------------
// The chaos contract: PRNG + pattern (mirror of core/c/fanout_chaos.c)
// ---------------------------------------------------------------------------

internal fun mix32(x0: UInt): UInt {
    var x = x0
    x = x xor (x shr 16)
    x = x * 0x7FEB352Du
    x = x xor (x shr 15)
    x = x * 0x846CA68Bu
    x = x xor (x shr 16)
    return x
}

internal class ChaosRng {
    var a: UInt = 0u; var b: UInt = 0u; var c: UInt = 0u; var d: UInt = 0u

    fun seed(seed: Int) {
        a = mix32(seed.toUInt() xor 0xA341316Cu)
        b = mix32(seed.toUInt() xor 0xC8013EA4u)
        c = a xor 0x9E3779B9u
        d = b xor 0x85EBCA6Bu
    }

    fun next(): UInt {
        var t = d
        val s = a
        d = c; c = b; b = s
        t = t xor (t shl 11)
        t = t xor (t shr 8)
        a = t xor s xor (s shr 10)
        return a
    }
}

internal fun tword(seq: Int, w: Int): UInt =
    mix32((seq.toUInt() * 2654435761u) + w.toUInt())

// ---------------------------------------------------------------------------
// Stepped engine — the deterministic scheduler (mirror of the C oracle)
// ---------------------------------------------------------------------------

private const val W_IDLE = 0; private const val W_FILL = 1
private const val W_STAMP = 2; private const val W_PUBLISH = 3; private const val W_DONE = 4
private const val R_IDLE = 0; private const val R_STAMP_B = 1; private const val R_COPY = 2
private const val R_STAMP_A = 3; private const val R_ACCEPT = 4; private const val R_TICK_END = 5
private const val R_SKIP_TICK = 6; private const val R_EXHAUSTED_TICK = 7; private const val R_DONE = 8

private val FREEZE_ADD = intArrayOf(1, 2, 3, 0) // preempt / stall / throttle / reorder

fun runSteppedChaos(cfg: ChaosConfig): ChaosVerdict {
    require(cfg.slots in 2..64) { "chaos: slots out of [2,64]" }
    require(cfg.words in 1..64) { "chaos: words out of [1,64]" }
    require(cfg.readers in 1..4) { "chaos: readers out of [1,4]" }
    require(cfg.chaosRate in 0..1000) { "chaos: chaosRate out of [0,1000]" }

    val M = cfg.slots; val W = cfg.words; val R = cfg.readers; val F = cfg.frames

    // Model ring (u64 counters stay < 2^53 in any reachable run).
    val slotSeq = LongArray(M)
    val payload = Array(M) { UIntArray(W) }
    val bracketOpen = BooleanArray(M)
    var latest = 0L
    var publishes = 0L

    // Writer SM
    var ws = W_IDLE
    var wSeq = 1L
    var wSlot = 0
    var wWord = 0
    var wRev = false

    // Reader SMs
    val rs = IntArray(R) { R_IDLE }
    val rLast = LongArray(R)
    val rTarget = Array(R) { LongArray(W) }
    val rL = LongArray(R)
    val rSlot = IntArray(R)
    val rWord = IntArray(R)
    val rAttempts = IntArray(R)
    val rRevActive = BooleanArray(R)

    // Scheduler
    val rng = ChaosRng()
    rng.seed(cfg.seed)
    val freeze = IntArray(R + 1)
    val revFlag = BooleanArray(R + 1)

    // Ledger
    val fresh = LongArray(R)
    val dropped = LongArray(R)
    val skips = LongArray(R)
    val exhausted = LongArray(R)
    var tornAccepted = 0L
    var futureClaims = 0L
    var bracketViolations = 0L
    val injected = LongArray(4)
    var stepsExecuted = 0L

    fun allDone(): Boolean {
        if (ws != W_DONE) return false
        for (i in 0 until R) if (rs[i] != R_DONE) return false
        return true
    }

    fun faultDraw() {
        if ((rng.next() % 1000u).toInt() >= cfg.chaosRate) return
        val victim = (rng.next() % (R.toUInt() + 1u)).toInt()
        val kind = (rng.next() % 4u).toInt()
        freeze[victim] += FREEZE_ADD[kind]
        if (kind == 3) revFlag[victim] = !revFlag[victim]
        injected[kind]++
    }

    fun writerStep() {
        when (ws) {
            W_IDLE -> {
                if (wSeq > F) { ws = W_DONE; return }
                wSlot = ((wSeq - 1) % M).toInt()
                if (bracketOpen[wSlot]) bracketViolations++ // L-C6
                slotSeq[wSlot] = 0                          // FI1a: invalidate FIRST
                bracketOpen[wSlot] = true
                wRev = revFlag[0]                           // fault axis: reorder
                wWord = if (wRev) W - 1 else 0
                ws = W_FILL
            }
            W_FILL -> {
                payload[wSlot][wWord] = tword(wSeq.toInt(), wWord)
                if (wRev) {
                    if (wWord == 0) ws = W_STAMP else wWord--
                } else {
                    wWord++
                    if (wWord == W) ws = W_STAMP
                }
            }
            W_STAMP -> {
                slotSeq[wSlot] = wSeq                       // FI1b: stamp (Release)
                bracketOpen[wSlot] = false
                ws = W_PUBLISH
            }
            W_PUBLISH -> {
                latest = wSeq                               // the publication point
                publishes++
                wSeq++
                ws = W_IDLE
            }
        }
    }

    fun readerStep(i: Int) {
        when (rs[i]) {
            R_IDLE -> {
                val L = latest
                if (L == 0L || L == rLast[i]) { rs[i] = R_TICK_END; return }
                rL[i] = L
                rAttempts[i] = 0
                rs[i] = R_STAMP_B
            }
            R_STAMP_B -> {
                val L = rL[i]
                rSlot[i] = ((L - 1) % M).toInt()
                val sB = slotSeq[rSlot[i]]
                if (sB != L) {
                    val l2 = latest
                    if (l2 == L) { skips[i]++; rs[i] = R_SKIP_TICK; return }
                    rL[i] = l2
                    rAttempts[i]++
                    rs[i] = if (rAttempts[i] >= 4) R_EXHAUSTED_TICK else R_STAMP_B
                    return
                }
                rWord[i] = if (revFlag[i + 1]) W - 1 else 0  // fault axis: reorder
                rRevActive[i] = revFlag[i + 1]
                rs[i] = R_COPY
            }
            R_COPY -> {
                val w = rWord[i]
                rTarget[i][w] = payload[rSlot[i]][w].toLong()
                if (rRevActive[i]) {
                    if (w == 0) rs[i] = R_STAMP_A else rWord[i] = w - 1
                } else {
                    rWord[i]++
                    if (rWord[i] == W) rs[i] = R_STAMP_A
                }
            }
            R_STAMP_A -> {
                val sA = slotSeq[rSlot[i]]
                if (sA == rL[i]) { rs[i] = R_ACCEPT; return }
                rL[i] = latest                               // torn copy — chase newest
                rAttempts[i]++
                rs[i] = if (rAttempts[i] >= 4) R_EXHAUSTED_TICK else R_STAMP_B
            }
            R_ACCEPT -> {
                val L = rL[i]
                if (L > F) futureClaims++                    // L-C2
                for (w in 0 until W) {
                    if (rTarget[i][w] != tword(L.toInt(), w).toLong()) {
                        tornAccepted++; break                // L-C1
                    }
                }
                dropped[i] += L - rLast[i] - 1
                rLast[i] = L
                fresh[i]++
                rs[i] = R_TICK_END
            }
            R_TICK_END -> {
                rs[i] = if (ws == W_DONE && rLast[i] == F.toLong()) R_DONE else R_IDLE
            }
            R_SKIP_TICK -> rs[i] = R_TICK_END
            R_EXHAUSTED_TICK -> { exhausted[i]++; rs[i] = R_TICK_END }
        }
    }

    fun advance(tid: Int) {
        if (tid == 0) writerStep() else readerStep(tid - 1)
    }

    // The scheduler loop (chaos contract shape — identical to the C oracle).
    while (stepsExecuted < cfg.steps && !allDone()) {
        val tid = (rng.next() % (R.toUInt() + 1u)).toInt()
        faultDraw()
        if (freeze[tid] > 0) freeze[tid]--   // scheduled, but made no progress
        else advance(tid)
        stepsExecuted++
    }

    // Drain: round-robin (writer, reader 1..R), no faults, until done.
    val drainBound = 64L * (F + 16L * R * (F + 8)) + 64
    var drainedSteps = 0L
    while (!allDone() && drainedSteps < drainBound) {
        for (tid in 0..R) {
            if (allDone()) break
            if (freeze[tid] > 0) freeze[tid] = 0  // faults end with the budget
            if (!allDone()) advance(tid)
        }
        drainedSteps++
    }
    val drained = allDone()

    // L-C3 telescoping + L-C4 completion (adjudicated once, at drain end).
    var telescopingOk = drained
    for (i in 0 until R) {
        if (dropped[i] != rLast[i] - fresh[i]) telescopingOk = false
        if (drained && rLast[i] != F.toLong()) telescopingOk = false
    }
    if (publishes != F.toLong()) telescopingOk = false // L-C4

    val pass = telescopingOk && tornAccepted == 0L && futureClaims == 0L &&
               bracketViolations == 0L

    // Contract JSON — byte-identical to the C oracle (fixed field order).
    val json = StringBuilder()
    json.append("{\"engine\":\"weft-chaos-stepped\",\"v\":1,\"seed\":${cfg.seed.toUInt()},")
    json.append("\"steps\":${cfg.steps},")
    json.append("\"slots\":$M,\"words\":$W,\"readers\":$R,\"frames\":$F,")
    json.append("\"chaosRate\":${cfg.chaosRate},")
    json.append("\"stepsExecuted\":$stepsExecuted,")
    json.append("\"injections\":{\"preempt\":${injected[0]},\"stall\":${injected[1]},")
    json.append("\"throttle\":${injected[2]},\"reorder\":${injected[3]}},")
    json.append("\"ledger\":{\"publishes\":$publishes,")
    json.append("\"fresh\":[${fresh.joinToString(",")}],")
    json.append("\"dropped\":[${dropped.joinToString(",")}],")
    json.append("\"lastSeq\":[${rLast.joinToString(",")}],")
    json.append("\"skips\":[${skips.joinToString(",")}],")
    json.append("\"exhausted\":[${exhausted.joinToString(",")}],")
    json.append("\"tornAccepted\":$tornAccepted,\"futureClaims\":$futureClaims,")
    json.append("\"bracketViolations\":$bracketViolations,\"drained\":$drained},")
    val telescopingStr = if (pass) "OK" else "VIOLATED"
    val verdictStr = if (pass) "PASS" else "FAIL"
    json.append("\"telescoping\":\"$telescopingStr\",")
    json.append("\"verdict\":\"$verdictStr\"}")

    return ChaosVerdict(pass, "stepped", json.toString())
}

// Pinned vectors — MUST match core/c/fanout_chaos.c's selftest (the parity
// anchor; see tools/chaos-fixtures/).
fun chaosSelftest(): Boolean {
    if (mix32(0xDEADBEEFu) != 3861431939u) return false
    if (tword(1, 0) != 1834104592u) return false
    if (tword(7, 3) != 2500287888u) return false
    if (tword(100, 15) != 4197121613u) return false
    val rng = ChaosRng()
    rng.seed(42)
    val draws = uintArrayOf(1034221180u, 2302191726u, 1921777443u, 3822789115u,
                            4193179225u, 3051818586u, 2559959645u, 2724063783u)
    for (want in draws) {
        if (rng.next() != want) return false
    }
    val v = runSteppedChaos(ChaosConfig(7, 4000, 2, 2, 2, 4, 300))
    return v.pass && v.json.contains("\"verdict\":\"PASS\"")
}

// CLI — stepped mode only (the parity oracle surface; the C runner carries
// the free-mode tier on real OS threads).
fun main(args: Array<String>) {
    if (args.size >= 8 && args[0] == "stepped") {
        val cfg = ChaosConfig(
            seed = args[7].toUInt().toInt(),
            steps = args[1].toLong(),
            slots = args[2].toInt(),
            words = args[3].toInt(),
            readers = args[4].toInt(),
            frames = args[5].toInt(),
            chaosRate = args[6].toInt(),
        )
        val v = runSteppedChaos(cfg)
        println(v.json)
        System.err.println(
            "chaos-stepped(jvm): ${if (v.pass) "PASS" else "FAIL"}")
        exitProcess(if (v.pass) 0 else 1)
    }
    if (args.size == 1 && args[0] == "selftest") {
        val ok = chaosSelftest()
        System.err.println("chaos-selftest(jvm): ${if (ok) "PASS" else "FAIL"}")
        exitProcess(if (ok) 0 else 1)
    }
    System.err.println("usage: FanoutChaos stepped <steps> <slots> <words> <readers> <frames> <chaosRate> <seed> | selftest")
    exitProcess(2)
}

File: core/dart/fanout_chaos.dart
// fanout_chaos.dart — RFC 0011 deterministic chaos engine, Dart port.
//
// "chaos contract" — mirrors core/c/fanout_chaos.c (the reference oracle)
// micro-step for micro-step, draw for draw. Given the same config, the
// stepped verdict JSON is BYTE-IDENTICAL to the C engine's (verified by
// ci/scripts/run_chaos_parity.sh; the golden fixture lives in
// tools/chaos-fixtures/). Read core/c/fanout_chaos.h for the normative
// contract; this port adds nothing and drops nothing.
//
// u32 discipline: Dart VM ints are 64-bit signed; every contract operation
// masks to the low 32 bits (`& 0xFFFFFFFF`) so the arithmetic is the exact
// analog of C's uint32_t. `>>>` (logical shift right) since Dart 2.14.
// tword's product exceeds 2^32 but stays < 2^63, so the low-32 mask after
// the multiply is exact (no precision loss) — same low-32 result as C.
//
// LAW 4 honesty: the stepped engine models the RFC 0004 protocol exactly
// (FI1/FI2 brackets, the bounded 4-attempt claim loop, latest-wins). The
// real-ring Flutter chaos leg lives in packages/flutter_weft tests; this
// file is the standalone parity oracle (no Flutter dependency).

import 'dart:io';

class ChaosConfig {
  final int seed; // master seed (u32)
  final int steps; // scheduler step budget (stepped mode)
  final int slots; // M — ring depth
  final int words; // W — payload words per slot
  final int readers; // R — concurrent reader SMs
  final int frames; // F — frames the writer publishes
  final int chaosRate; // per-mille fault probability per step

  const ChaosConfig(this.seed, this.steps, this.slots, this.words,
      this.readers, this.frames, this.chaosRate);
}

class ChaosVerdict {
  final bool pass;
  final String engine;
  final String json; // the contract JSON (byte-identical across ports)

  const ChaosVerdict(this.pass, this.engine, this.json);
}

// ---------------------------------------------------------------------------
// The chaos contract: PRNG + pattern (mirror of core/c/fanout_chaos.c)
// ---------------------------------------------------------------------------

const int _u32Mask = 0xFFFFFFFF;

int _mix32(int x0) {
  var x = x0 & _u32Mask;
  x = (x ^ (x >>> 16)) & _u32Mask;
  x = (x * 0x7FEB352D) & _u32Mask;
  x = (x ^ (x >>> 15)) & _u32Mask;
  x = (x * 0x846CA68B) & _u32Mask;
  x = (x ^ (x >>> 16)) & _u32Mask;
  return x;
}

class _ChaosRng {
  int _a = 0, _b = 0, _c = 0, _d = 0;

  void seed(int seed) {
    _a = _mix32(seed ^ 0xA341316C);
    _b = _mix32(seed ^ 0xC8013EA4);
    _c = (_a ^ 0x9E3779B9) & _u32Mask;
    _d = (_b ^ 0x85EBCA6B) & _u32Mask;
  }

  int next() {
    var t = _d;
    final s = _a;
    _d = _c;
    _c = _b;
    _b = s;
    t = (t ^ ((t << 11) & _u32Mask)) & _u32Mask;
    t = (t ^ (t >>> 8)) & _u32Mask;
    _a = (t ^ s ^ (s >>> 10)) & _u32Mask;
    return _a;
  }
}

int _tword(int seq, int w) =>
    _mix32(((seq * 2654435761) + w) & _u32Mask);

// ---------------------------------------------------------------------------
// Stepped engine — the deterministic scheduler (mirror of the C oracle)
// ---------------------------------------------------------------------------

const int _wIdle = 0, _wFill = 1, _wStamp = 2, _wPublish = 3, _wDone = 4;
const int _rIdle = 0, _rStampB = 1, _rCopy = 2, _rStampA = 3, _rAccept = 4;
const int _rTickEnd = 5, _rSkipTick = 6, _rExhaustedTick = 7, _rDone = 8;

const List<int> _freezeAdd = [1, 2, 3, 0]; // preempt/stall/throttle/reorder

ChaosVerdict runSteppedChaos(ChaosConfig cfg) {
  if (cfg.slots < 2 || cfg.slots > 64) {
    throw ArgumentError('chaos: slots out of [2,64]');
  }
  if (cfg.words < 1 || cfg.words > 64) {
    throw ArgumentError('chaos: words out of [1,64]');
  }
  if (cfg.readers < 1 || cfg.readers > 4) {
    throw ArgumentError('chaos: readers out of [1,4]');
  }
  if (cfg.chaosRate < 0 || cfg.chaosRate > 1000) {
    throw ArgumentError('chaos: chaosRate out of [0,1000]');
  }

  final M = cfg.slots, W = cfg.words, R = cfg.readers, F = cfg.frames;

  // Model ring.
  final slotSeq = List<int>.filled(M, 0);
  final payload = List.generate(M, (_) => List<int>.filled(W, 0));
  final bracketOpen = List<bool>.filled(M, false);
  var latest = 0;
  var publishes = 0;

  // Writer SM
  var ws = _wIdle;
  var wSeq = 1;
  var wSlot = 0;
  var wWord = 0;
  var wRev = false;

  // Reader SMs
  final rs = List<int>.filled(R, _rIdle);
  final rLast = List<int>.filled(R, 0);
  final rTarget = List.generate(R, (_) => List<int>.filled(W, 0));
  final rL = List<int>.filled(R, 0);
  final rSlot = List<int>.filled(R, 0);
  final rWord = List<int>.filled(R, 0);
  final rAttempts = List<int>.filled(R, 0);
  final rRevActive = List<bool>.filled(R, false);

  // Scheduler
  final rng = _ChaosRng()..seed(cfg.seed);
  final freeze = List<int>.filled(R + 1, 0);
  final revFlag = List<bool>.filled(R + 1, false);

  // Ledger
  final fresh = List<int>.filled(R, 0);
  final dropped = List<int>.filled(R, 0);
  final skips = List<int>.filled(R, 0);
  final exhausted = List<int>.filled(R, 0);
  var tornAccepted = 0;
  var futureClaims = 0;
  var bracketViolations = 0;
  final injected = List<int>.filled(4, 0);
  var stepsExecuted = 0;

  bool allDone() {
    if (ws != _wDone) return false;
    for (var i = 0; i < R; i++) {
      if (rs[i] != _rDone) return false;
    }
    return true;
  }

  void faultDraw() {
    if (rng.next() % 1000 >= cfg.chaosRate) return;
    final victim = rng.next() % (R + 1);
    final kind = rng.next() % 4;
    freeze[victim] += _freezeAdd[kind];
    if (kind == 3) revFlag[victim] = !revFlag[victim];
    injected[kind]++;
  }

  void writerStep() {
    switch (ws) {
      case _wIdle:
        if (wSeq > F) {
          ws = _wDone;
          return;
        }
        wSlot = (wSeq - 1) % M;
        if (bracketOpen[wSlot]) bracketViolations++; // L-C6
        slotSeq[wSlot] = 0; // FI1a: invalidate FIRST
        bracketOpen[wSlot] = true;
        wRev = revFlag[0]; // fault axis: reorder
        wWord = wRev ? (W - 1) : 0;
        ws = _wFill;
      case _wFill:
        payload[wSlot][wWord] = _tword(wSeq, wWord);
        if (wRev) {
          if (wWord == 0) {
            ws = _wStamp;
          } else {
            wWord--;
          }
        } else {
          wWord++;
          if (wWord == W) ws = _wStamp;
        }
      case _wStamp:
        slotSeq[wSlot] = wSeq; // FI1b: stamp (Release)
        bracketOpen[wSlot] = false;
        ws = _wPublish;
      case _wPublish:
        latest = wSeq; // the publication point
        publishes++;
        wSeq++;
        ws = _wIdle;
    }
  }

  void readerStep(int i) {
    switch (rs[i]) {
      case _rIdle:
        final L = latest;
        if (L == 0 || L == rLast[i]) {
          rs[i] = _rTickEnd;
          return;
        }
        rL[i] = L;
        rAttempts[i] = 0;
        rs[i] = _rStampB;
      case _rStampB:
        final L = rL[i];
        rSlot[i] = (L - 1) % M;
        final sB = slotSeq[rSlot[i]];
        if (sB != L) {
          final l2 = latest;
          if (l2 == L) {
            skips[i]++;
            rs[i] = _rSkipTick;
            return;
          }
          rL[i] = l2;
          rAttempts[i]++;
          rs[i] = rAttempts[i] >= 4 ? _rExhaustedTick : _rStampB;
          return;
        }
        rWord[i] = revFlag[i + 1] ? (W - 1) : 0; // fault axis: reorder
        rRevActive[i] = revFlag[i + 1];
        rs[i] = _rCopy;
      case _rCopy:
        final w = rWord[i];
        rTarget[i][w] = payload[rSlot[i]][w];
        if (rRevActive[i]) {
          if (w == 0) {
            rs[i] = _rStampA;
          } else {
            rWord[i] = w - 1;
          }
        } else {
          rWord[i]++;
          if (rWord[i] == W) rs[i] = _rStampA;
        }
      case _rStampA:
        final sA = slotSeq[rSlot[i]];
        if (sA == rL[i]) {
          rs[i] = _rAccept;
          return;
        }
        rL[i] = latest; // torn copy — chase newest
        rAttempts[i]++;
        rs[i] = rAttempts[i] >= 4 ? _rExhaustedTick : _rStampB;
      case _rAccept:
        final L = rL[i];
        if (L > F) futureClaims++; // L-C2
        for (var w = 0; w < W; w++) {
          if (rTarget[i][w] != _tword(L, w)) {
            tornAccepted++; // L-C1
            break;
          }
        }
        dropped[i] += L - rLast[i] - 1;
        rLast[i] = L;
        fresh[i]++;
        rs[i] = _rTickEnd;
      case _rTickEnd:
        rs[i] = (ws == _wDone && rLast[i] == F) ? _rDone : _rIdle;
      case _rSkipTick:
        rs[i] = _rTickEnd;
      case _rExhaustedTick:
        exhausted[i]++;
        rs[i] = _rTickEnd;
    }
  }

  void advance(int tid) {
    if (tid == 0) {
      writerStep();
    } else {
      readerStep(tid - 1);
    }
  }

  // The scheduler loop (chaos contract shape — identical to the C oracle).
  while (stepsExecuted < cfg.steps && !allDone()) {
    final tid = rng.next() % (R + 1);
    faultDraw();
    if (freeze[tid] > 0) {
      freeze[tid]--; // scheduled, but made no progress
    } else {
      advance(tid);
    }
    stepsExecuted++;
  }

  // Drain: round-robin (writer, reader 1..R), no faults, until done.
  final drainBound = 64 * (F + 16 * R * (F + 8)) + 64;
  var drainedSteps = 0;
  while (!allDone() && drainedSteps < drainBound) {
    for (var tid = 0; tid <= R && !allDone(); tid++) {
      if (freeze[tid] > 0) freeze[tid] = 0; // faults end with the budget
      if (!allDone()) advance(tid);
    }
    drainedSteps++;
  }
  final drained = allDone();

  // L-C3 telescoping + L-C4 completion (adjudicated once, at drain end).
  var telescopingOk = drained;
  for (var i = 0; i < R; i++) {
    if (dropped[i] != rLast[i] - fresh[i]) telescopingOk = false;
    if (drained && rLast[i] != F) telescopingOk = false;
  }
  if (publishes != F) telescopingOk = false; // L-C4

  final pass = telescopingOk &&
      tornAccepted == 0 &&
      futureClaims == 0 &&
      bracketViolations == 0;

  // Contract JSON — byte-identical to the C oracle (fixed field order).
  final json = '{"engine":"weft-chaos-stepped","v":1,"seed":${cfg.seed},'
      '"steps":${cfg.steps},'
      '"slots":$M,"words":$W,"readers":$R,"frames":$F,'
      '"chaosRate":${cfg.chaosRate},'
      '"stepsExecuted":$stepsExecuted,'
      '"injections":{"preempt":${injected[0]},"stall":${injected[1]},'
      '"throttle":${injected[2]},"reorder":${injected[3]}},'
      '"ledger":{"publishes":$publishes,'
      '"fresh":[${fresh.join(",")}],'
      '"dropped":[${dropped.join(",")}],'
      '"lastSeq":[${rLast.join(",")}],'
      '"skips":[${skips.join(",")}],'
      '"exhausted":[${exhausted.join(",")}],'
      '"tornAccepted":$tornAccepted,"futureClaims":$futureClaims,'
      '"bracketViolations":$bracketViolations,"drained":$drained},'
      '"telescoping":"${pass ? "OK" : "VIOLATED"}",'
      '"verdict":"${pass ? "PASS" : "FAIL"}"}';

  return ChaosVerdict(pass, 'stepped', json);
}

// Pinned vectors — MUST match core/c/fanout_chaos.c's selftest (the parity
// anchor; see tools/chaos-fixtures/).
bool chaosSelftest() {
  if (_mix32(0xDEADBEEF) != 3861431939) return false;
  if (_tword(1, 0) != 1834104592) return false;
  if (_tword(7, 3) != 2500287888) return false;
  if (_tword(100, 15) != 4197121613) return false;
  final rng = _ChaosRng()..seed(42);
  const draws = [1034221180, 2302191726, 1921777443, 3822789115, 4193179225,
    3051818586, 2559959645, 2724063783];
  for (final want in draws) {
    if (rng.next() != want) return false;
  }
  final v = runSteppedChaos(
      const ChaosConfig(7, 4000, 2, 2, 2, 4, 300));
  return v.pass && v.json.contains('"verdict":"PASS"');
}

void main(List<String> args) {
  if (args.length >= 8 && args[0] == 'stepped') {
    final cfg = ChaosConfig(
      int.parse(args[7]), // seed
      int.parse(args[1]), // steps
      int.parse(args[2]), // slots
      int.parse(args[3]), // words
      int.parse(args[4]), // readers
      int.parse(args[5]), // frames
      int.parse(args[6]), // chaosRate
    );
    final v = runSteppedChaos(cfg);
    print(v.json);
    stderr.writeln('chaos-stepped(dart): ${v.pass ? "PASS" : "FAIL"}');
    exit(v.pass ? 0 : 1);
  }
  if (args.length == 1 && args[0] == 'selftest') {
    final ok = chaosSelftest();
    stderr.writeln('chaos-selftest(dart): ${ok ? "PASS" : "FAIL"}');
    exit(ok ? 0 : 1);
  }
  stderr.writeln(
      'usage: fanout_chaos.dart stepped <steps> <slots> <words> <readers> '
      '<frames> <chaosRate> <seed> | selftest');
  exit(2);
}
